features/bugs/go/activity_start_race/feature.go

```diff
@@ -0,0 +1,138 @@
+//go:build !pre1.12.0
+
+package activity_start_race
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	"go.temporal.io/sdk-features/harness/go/harness"
+	"go.temporal.io/sdk-features/harness/go/history"
+	"go.temporal.io/sdk/client"
+	"go.temporal.io/sdk/workflow"
+	"google.golang.org/grpc"
+)
+
+var Feature = harness.Feature{
+	Workflows:  Workflow,
+	Activities: SleepActivity,
+	Execute:    Execute,
+	ClientOptions: client.Options{
+		ConnectionOptions: client.ConnectionOptions{
+			DialOptions: []grpc.DialOption{grpc.WithUnaryInterceptor(delayWorkflow)},
+		},
+	},
+	CheckHistory: CheckHistory,
+}
+
+var pollWorkflowWait, pollActivityWait sync.WaitGroup
+
+const activitySleep = 200 * time.Millisecond
+
+func Execute(ctx context.Context, r *harness.Runner) (client.WorkflowRun, error) {
+	// Make the next activity poll wait
+	pollActivityWait.Add(1)
+
+	// Start the workflow
+	run, err := r.ExecuteDefault(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make workflow poll wait
+	pollWorkflowWait.Add(1)
+
+	// Resume workflow
+	r.Require.NoError(r.Client.SignalWorkflow(ctx, run.GetID(), run.GetRunID(), "resume", nil))
+
+	// Wait a bit, let activity poll succeed
+	time.Sleep(activitySleep * 2)
+	pollActivityWait.Done()
+
+	// Wait a bit, then let workflow poll succeed
+	time.Sleep(activitySleep * 4)
+	pollWorkflowWait.Done()
+
+	// Wait a bit, then cancel the workflow
+	time.Sleep(activitySleep * 2)
+	r.Require.NoError(r.Client.CancelWorkflow(ctx, run.GetID(), run.GetRunID()))
+	return run, err
+}
+
+func CheckHistory(ctx context.Context, r *harness.Runner, run client.WorkflowRun) error {
+	// We want to do the default history check which just does a replay of the
```

Comment:
Can't we make this completely deterministic?
And if you're using manually generated history, do we need this complexity in the feature? Are you just keeping it around for future use-cases where we need to try and reproduce timing sensitive histories?

Reply:
> Can't we make this completely deterministic?

I cannot make it completely deterministic I don't think without slowing things way down. My attempts to make it reliably return the same order of events failed, I can only get it to do so _sometimes_.

>  And if you're using manually generated history, do we need this complexity in the feature? Are you just keeping it around for future use-cases where we need to try and reproduce timing sensitive histories?

I actually generated this history, it just doesn't happen in this order every time. I keep it around for this specific replay. So yeah, there's kinda two separate tests here: 1) that the workflow runs as expected and 2) that the workflow works with the given history. The history of #1 is not checked/used.